File: leetcode/67.cpp
#include<cstring>
using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
        int aindex=a.length();
        int bindex=b.length();
        int ansindex=(aindex>bindex?aindex:bindex)+1;
        string ans(ansindex, '0');

        while(--ansindex>=0){
            int anum=--aindex>=0?a[aindex]-'0':0;
            int bnum=--bindex>=0?b[bindex]-'0':0;
            int carry=ans[ansindex]-'0';
            ans[ansindex]=(anum^bnum^carry)+'0';
            if(ansindex) ans[ansindex-1]=(anum&bnum|bnum&carry|carry&anum)+'0';
        }
        
        return ans[0]=='0'?ans.substr(1):ans;
    }
};

File: leetcode/67.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        x, y=int(a, 2), int(b, 2)
        while y:
            x, y=x^y, (x&y)<<1
        return bin(x)[2:]

#int(str, sys) 将字符串转为sys进制的数字 >>>int('18', 16)  :24
#bin() 将整数转为二进制表示 >>>bin(23)  :0a11000  [2:]的作用是切片 将0a去除